db/census.go

```diff
@@ -58,6 +59,97 @@ func (ms *MongoStorage) SetCensus(census *Census) (string, error) {
 	return census.ID.Hex(), nil
 }
 
+// SetPublished census updates the PublishedCensus field of a census
+func (ms *MongoStorage) SetPublishedCensus(censusID, uri string, root internal.HexBytes) (string, error) {
+	if len(censusID) == 0 || len(uri) == 0 || len(root) == 0 {
+		return "", ErrInvalidData
+	}
+
+	censusOID, err := primitive.ObjectIDFromHex(censusID)
+	if err != nil {
+		return "", ErrInvalidData
+	}
+	census := &Census{
+		ID: censusOID,
+		Published: PublishedCensus{
+			Root: root,
+			URI:  uri,
+		},
+	}
+
+	return ms.SetCensus(census)
+}
+
+// SetGroupCensus creates a new census for an organization
+// Returns the hex representation of the census
+func (ms *MongoStorage) SetGroupCensus(
+	census *Census,
+	groupID string,
+	participantIDs []primitive.ObjectID,
+) (string, error) {
+	if len(groupID) == 0 {
+		return ms.SetCensus(census)
+	}
+
+	// create a context with a timeout
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+
+	if census.OrgAddress == "" {
+		return "", ErrInvalidData
+	}
+
+	// check that the org exists
+	_, err := ms.Organization(census.OrgAddress)
+	if err != nil {
+		if err == ErrNotFound {
+			return "", ErrInvalidData
+		}
+		return "", fmt.Errorf("error retrieving organization: %w", err)
+	}
+
+	// check that the group exists
+	group, err := ms.OrganizationMemberGroup(groupID, census.OrgAddress)
+	if err != nil {
+		if err == ErrNotFound {
+			return "", ErrInvalidData
+		}
+		return "", fmt.Errorf("error retrieving organization group: %w", err)
+	}
+	census.GroupID = group.ID
+
+	if census.ID != primitive.NilObjectID {
+		// if the census exists, update it with the new data
+		census.UpdatedAt = time.Now()
+	} else {
+		// if the census doesn't exist, create its id
+		census.ID = primitive.NewObjectID()
+		census.CreatedAt = time.Now()
+	}
+
+	updateDoc, err := dynamicUpdateDocument(census, nil)
+	if err != nil {
+		return "", err
+	}
+	ms.keysLock.Lock()
+	defer ms.keysLock.Unlock()
```

Comment:
this lock should happen before `ms.Organization(census.OrgAddress)` to achieve atomic db operations across different documents